### PR TITLE
Add `remove_duplicates` ignore_comments and merge_comments logic

### PR DESCRIPTION
**Pull request by jakos-sec**
Added `DuplicateResolution` enum to handle advanced deduplication modes. The new mores are `ignore_comments` to keep only the first comment for deduplicated items and `merge_comments` to merge comments from all identical items when deduplicated (identical comment lines are also deduplicated when merging).

### CODE REVIEW

File: keepsorted/block.go [hunk 1]
```diff
@@ -259,15 +259,35 @@ func (b block) sorted() (sorted []string, alreadySorted bool) {
...
+				if firstLg, ok := seenLines[codeMapKey]; !ok {
+					seenLines[codeMapKey] = lg
+					deduped = append(deduped, lg)
+				} else {
+					removedDuplicate = true
+
+					if b.metadata.opts.RemoveDuplicates == DuplicateResolutionMergeComments {
+						for _, newComment := range lg.comment {
+							if !slices.Contains(firstLg.comment, newComment) {
+								firstLg.comment = append(firstLg.comment[:len(firstLg.comment):len(firstLg.comment)], newComment)
```
Comment (wachsler-google) [inline keepsorted/block.go]: The 3-index slice here is defensive against a shared backing array, which makes sense — but it's pretty hard to read, and it also forces a new allocation on every iteration of the loop rather than just once.

Consider using `slices.Clone` before the loop instead, which communicates the intent more clearly and lets `append` do its normal amortized growth:

```go
firstLg.comment = slices.Clone(firstLg.comment)
for _, newComment := range lg.comment {
    if !slices.Contains(firstLg.comment, newComment) {
        firstLg.comment = append(firstLg.comment, newComment)
    }
}
```

File: keepsorted/options.go [hunk 2]
```diff
@@ -35,6 +35,16 @@ import (
...
 
+type DuplicateResolution int
+
+const (
+	DuplicateResolutionFalse DuplicateResolution = iota
+	DuplicateResolutionTrue
+	DuplicateResolutionIgnoreComments
+	DuplicateResolutionMergeComments
+)
+
```
Comment (wachsler-google) [inline keepsorted/options.go:53]: nit: extra blank line

File: keepsorted/options_parser.go [hunk 3]
```diff
@@ -98,6 +101,23 @@ func (p *parser) popBool() (bool, error) {
...
 
+func (p *parser) popDuplicateResolution() (DuplicateResolution, error) {
+	val, rest, _ := strings.Cut(p.line, " ")
+	p.line = rest
+	switch val {
+	case "yes", "true":
+		return DuplicateResolutionTrue, nil
+	case "no", "false":
+		return DuplicateResolutionFalse, nil
+	case "keep_first_comment", "ignore_comments":
```
Comment (wachsler-google) [inline keepsorted/options_parser.go]: This allows "keep_first_comment" as an alias for "ignore_comments", but it's not documented or tested. Per Hyrum's Law, let's not include an undocumented API surface.

IMHO "keep_first_comment" is clearer than "ignore_comments", so maybe just change to "keep_first_comment" everywhere?

**Review by wachsler-google — DISMISSED** [review]
Sorry I don't have much experience with github reviews, so I posted a couple comments separate from this review before I figured out how this works. Anyway, this all looks pretty good, aside from the few things I've commented on.

File: keepsorted/block.go (hunk 1 above)
Comment (jakos-sec) [reply keepsorted/block.go]: Makes sense, thank you! Fixed.

File: keepsorted/options_parser.go (hunk 3 above)
Comment (jakos-sec) [reply keepsorted/options_parser.go]: I agree, "keep_first_comment" is more descriptive. Changed it everywhere and removed "ignore_comments".

**Comment by jakos-sec** [comment]
> Sorry I don't have much experience with github reviews, so I posted a couple comments separate from this review before I figured out how this works. Anyway, this all looks pretty good, aside from the few things I've commented on.

Feedback should be all addressed now. Thank you for the review!

File: keepsorted/block.go [hunk 4]
```diff
@@ -259,15 +259,36 @@ func (b block) sorted() (sorted []string, alreadySorted bool) {
...
+
+				if firstLg, ok := seenLines[codeMapKey]; !ok {
+					seenLines[codeMapKey] = lg
+					deduped = append(deduped, lg)
+				} else {
+					removedDuplicate = true
+
+					if b.metadata.opts.RemoveDuplicates == DuplicateResolutionMergeComments {
+						firstLg.comment = slices.Clone(firstLg.comment)
+						for _, newComment := range lg.comment {
```
Comment (KatrinaHoffert) [inline keepsorted/block.go]: I think you should not be looping over lines and doing a contains check. Doesn't that backfire if there are common lines? Particularly, I think this would strip all subsequent paragraph breaks.

I think this deduping should only ignore comments where all the lines exist in that order (i.e., join the comment strings first).

Can we also add a test case for more complex, multi-line comments, to ensure that removing duplicates correctly handles paragraph breaks and doesn't get confused by similar lines? e.g., like:

```
// Foo is a solution to all problems.
//
// - It notably does both everything and nothing at the same time.
// - For details on what it does, see
// http://example.com/some-long-url-that-takes-up-the-whole-line
foo

// Foo is the most important value in this list.
// 
// A record of past conflicts stemming from this value is available at
// http://example.com/some-long-url-that-takes-up-the-whole-line
foo
```

Which I think should just result in the following naive appending, since I see those comments to *not* be duplicates even though they contain individual lines that technically are duplicates.

```
// Foo is a solution to all problems.
//
// - It notably does both everything and nothing at the same time.
// - For details on what it does, see
// http://example.com/some-long-url-that-takes-up-the-whole-line
// Foo is the most important value in this list.
// 
// A record of past conflicts stemming from this value is available at
// http://example.com/some-long-url-that-takes-up-the-whole-line
foo
```

I didn't run it to check, but I think your code would result in:

```
// Foo is a solution to all problems.
//
// - It notably does both everything and nothing at the same time.
// - For details on what it does, see
// http://example.com/some-long-url-that-takes-up-the-whole-line
// Foo is the most important value in this list.
// A record of past conflicts stemming from this value is available at
foo
```

Which IMO is a more confusing result. There's no achieving perfection, but I think merging at a line level rather than the entire comment is just going to confuse.

File: goldens/duplicates_keep_first_comment.out [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+// keep-sorted-test start remove_duplicates=keep_first_comment sticky_comments=yes
+bar
+// First foo
+foo
+// keep-sorted-test end
+
+Remove duplicates ignoring comments, when the first item has no comment:
+// keep-sorted-test start remove_duplicates=keep_first_comment sticky_comments=yes
+bar
+foo
```
Comment (KatrinaHoffert) [inline goldens/duplicates_keep_first_comment.out:12]: Is this behavior definitely what you want? I'm just not clear on the use case for it. Keeping the first actual comment seems like it would make more sense than letting the no-comment overwrite the duplicate that does have a comment.

WDYT if we change it so that keep_first_comment mode will check if the comment is empty for the first line group and if so, loop over all the line groups until it finds one with a non-empty comment (if any) and copies that comment over to the first line group?

File: keepsorted/block.go (hunk 4 above)
Comment (jakos-sec) [reply keepsorted/block.go]: I think that makes sense yes.

It might not be the most ideal for the use case we are looking at (the comments are build targets associated with entries in the ignorelist and we don't want those to duplicate if one of them is on multiple entries) but overall the better solution.

I don't think it will affect us or we can easily filter those out afterwards if it's an issue.

I changed it now as you suggested.

File: goldens/duplicates_keep_first_comment.out (hunk 5 above)
Comment (jakos-sec) [reply goldens/duplicates_keep_first_comment.out:12]: Good point, I've fixed the code to now take the first non-empty comment (if there is any).
